### DISCUSSION TITLE
Conditionnal import of the library

### DISCUSSION THREAD

**Discussion by RomainFALLER** [Arnaud73/ngx-matomo#102] [Q&A]
Hello, 

I would like to import the MatomoModule only if the user setted matomo configuration (conditionnally). I tried this 2 solutions but it doesn't work :

```  if (conf.matomoLocalConfiguration) {
      import('ngx-matomo').then((module) =>
      module.MatomoModule.forRoot({
        scriptUrl: conf.matomoLocalConfiguration.trackerUrl + 'matomo.js',
        trackers: [
          {
            trackerUrl: conf.matomoLocalConfiguration.trackerUrl + 'matomo.php',
            siteId: conf.matomoLocalConfiguration.idSite
          }
        ],
        routeTracking: {
          enable: true
        }
      }))
```
or this : 
```
		if (conf.matomoLocalConfiguration) {	
 				providers.push(MatomoModule.forRoot({
					scriptUrl: conf.matomoLocalConfiguration.trackerUrl + 'matomo.js',
					trackers: [
						{
							trackerUrl: conf.matomoLocalConfiguration.trackerUrl + 'matomo.php',
							siteId: conf.matomoLocalConfiguration.idSite
						}
					],
					routeTracking: {
						enable: true
					}
				}).providers!)
		}
		return {
			ngModule: AstModule,
			providers: [providers],
		};
	}
```

Also, I know you can call the MatomoModule.forRoot() in the imports and inject MATOMO_CONFIGURATION later, but the module is still imported even if the user don't need matomo. 

Does the library have another way of doing this? 

Best regards, 

Romain

**Comment by Arnaud73** [comment]
I guess your issue is related to providing async values for a module configuration. This problem is still being discussed here: https://github.com/angular/angular/issues/23279. One Angular developer provided as very sexy solution (https://github.com/angular/angular/issues/23279#issuecomment-1165030809) that does not work with modules since they are loaded too early in the process.

Regarding your issue, if providing the configuration as a Promise could be a solution (from what I understand), then check the branch `feature/configuration-as-promise` of ngx-matomo:
1. Clone the repo
2. Checkout the `feature/configuration-as-promise` branch
3. Have a look at the code of the demo app (it is using a Promise for the configuration)
4. Run it: `npm run demo`

This branch has not been integrated into main yet, because:
* it makes to code more complex and as a result everyone is going to pay for a heavier library. So I am still not convinced in merging it.
* it does not consider some options that will appear in the configuration and that would require tracking before the configuration is provided

If you have a look at it, let me know if it solves your problem!

**Reply by RomainFALLER** [reply]
Thank you for your answer. I will try it tomorrow and let you know.  :)

When I see the demo on the branch feature/configuration-as-promise, I understand that if user did not set configuration, MatomoModule will still be imported and the function MatomoModule.forRoot() will be called without parameter, that's right ?

Best regards,

Romain

**Comment by RomainFALLER** [comment]
Hello, thank you again for your answer. The promise allow us to configure conditionnally the matomo module even if I prefer this solution to do it :
```
...
imports: [
MatomoModule.forRoot(), // Call empry forRoot method
]
...
if (conf.matomoLocalConfiguration) {

     // Inject Matomo conf later if it is setted with the provided injection token MATOMO_CONFIGURATION
    providers.push({provide: MATOMO_CONFIGURATION, useValue: {
					scriptUrl: conf.matomoLocalConfiguration.trackerUrl + 'matomo.js',
					trackers: [
						{
							trackerUrl: conf.matomoLocalConfiguration.trackerUrl + 'matomo.php',
							siteId: conf.matomoLocalConfiguration.idSite
						}
					],
					routeTracking: {
						enable: true
					}
				}
                           });
}
...
```

Whether it's token injection or promise, I'm always obliged to import MatomoModule even if the user doesn't need it...

Do you see another way without importing automatically MatomoModule ? 

I can do this kind of thing with module like Msal for example.

Best regards,

Romain
